pennylane/transforms/diagonalize_measurements.py

```diff
@@ -121,9 +126,112 @@ def circuit(x):
             f"but received {list(bad_obs_input)}"
         )
 
+    diagonalize_all = set(supported_base_obs).issubset(set(_default_supported_obs))
+
+    if to_eigvals and not diagonalize_all:
+        raise ValueError(
+            f"Using to_eigvals=True requires diagonalizing all observables to the measurement basis. Observables {set(supported_base_obs)-set(_default_supported_obs)} can't be supported when using eigvals."
+        )
+
+    if (
+        all(m.obs.pauli_rep is not None for m in tape.measurements if m.obs is not None)
+        and diagonalize_all
+    ):
+        try:
+            if tape.samples_computational_basis and len(tape.measurements) > 1:
+                _validate_computational_basis_sampling(tape)
+            diagonalizing_gates, new_measurements = _diagonalize_all_pauli_obs(
+                tape, to_eigvals=to_eigvals
+            )
+        except qml.QuantumFunctionError:
+            # the pauli_rep based method sometimes fails unnecessarily -
+            # if it fails, fall back on the less efficient method (which may also fail)
```

Comment:
XD
Do we know why does it fail? It's just a curiosity at this stage as I am still looking at the [background](https://app.shortcut.com/xanaduai/story/60179/create-a-transform-that-decomposes-non-supported-observables-in-catalyst) and the previous PR #5829 (still learning about the context)

Reply:
Its possible I'm forgetting other edge cases right now, but the main thing would be if we had, as an example, `qml.expval(X(0)), qml.var(X(0)@Y(1)`, the pauli_rep method would just see two different operators, both on X(0), and raise an error. 

The fallback method iterates through each operator (which is why its less efficient), so it identifies that they are both X(0), and diagonalizes X(0) once.